sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs

```diff
@@ -1,79 +1,125 @@
             }
-
-            var stringBuilder = new StringBuilder();
-            stringBuilder.Append(eventData.EventName);
-
-            if (!string.IsNullOrWhiteSpace(eventData.Message))
+            catch (FormatException)
             {
-                stringBuilder.AppendLine();
-                stringBuilder.Append(nameof(eventData.Message)).Append(" = ").Append(eventData.Message);
             }
+        }
 
-            if (eventData.PayloadNames != null)
-            {
-                for (int i = 0; i < eventData.PayloadNames.Count; i++)
-                {
-                    stringBuilder.AppendLine();
-                    stringBuilder.Append(eventData.PayloadNames[i]).Append(" = ").Append(payloadArray[i]);
-                }
-            }
+        StringBuilder stringBuilder = RentStringBuilder();
+        stringBuilder.Append(eventData.EventName);
 
-            return stringBuilder.ToString();
+        if (!string.IsNullOrWhiteSpace(eventData.Message))
+        {
+            stringBuilder.AppendLine();
+            stringBuilder.Append(nameof(eventData.Message)).Append(" = ").Append(eventData.Message);
         }
 
-        private static void ProcessPayloadArray(object?[] payloadArray)
+        if (eventData.PayloadNames != null)
         {
-            for (int i = 0; i < payloadArray.Length; i++)
+            for (int i = 0; i < eventData.PayloadNames.Count; i++)
             {
-                payloadArray[i] = FormatValue(payloadArray[i]);
+                stringBuilder.AppendLine();
+                stringBuilder.Append(eventData.PayloadNames[i]).Append(" = ").Append(payloadArray[i]);
             }
         }
 
-        private static object? FormatValue(object? o)
+        return ToStringAndReturnStringBuilder(stringBuilder);
+    }
+
+    private static void ProcessPayloadArray(object?[] payloadArray)
+    {
+        for (int i = 0; i < payloadArray.Length; i++)
         {
-            if (o is byte[] bytes)
+            payloadArray[i] = FormatValue(payloadArray[i]);
+        }
+    }
+
+    private static object? FormatValue(object? o)
+    {
+        if (o is byte[] bytes)
+        {
+#if NET6_0_OR_GREATER
+            return Convert.ToHexString(bytes);
+#else
+            int length = 2 * bytes.Length; // Two hex characters per byte
+            if (length < 256)
             {
-                var stringBuilder = new StringBuilder();
-                foreach (byte b in bytes)
+                static char ToHex(int value) => (char)(value < 10 ? value + '0' : value - 10 + 'A');
+
+                Span<char> buffer = stackalloc char[length];
+                for (int i = 0; i < bytes.Length; i++)
                 {
-                    stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:X2}", b);
+                    byte b = bytes[i];
+                    buffer[i * 2] = ToHex(b >> 4); // Shift upper nibble into lower 4 bits and convert to hex
```

Comment:
What I'd like to capture here is why.   Why do we want to shift before converting?  What's the overall algorithm/approach that we're implementing?  

Our goal is to be able to have our team as a whole maintain and contribute to Core, regardless of level or years of experience.   The previous version of the code was less efficient but more approachable.  While I agree this is a path worth optimizing, I want to be sure that we don't add complexity without making sure that it can be easily understood by all.

Reply:
I appreciate the feedback, thank you. What I've done in the latest iteration is taken the algorithm from the .NET runtime. It's even faster than I had before. The algorithm is open source (the same MIT license as this repo) and heavily commented so, hopefully it can be understood by everyone on the team.

Reply:
Much appreciated.  The comments pointing towards the inspiration are grea.